test/cql-pytest/test_batch.py

```diff
@@ -7,9 +7,9 @@
 # Tests for batch operations
 #############################################################################
 from cassandra import InvalidRequest
-
+from cassandra.cluster import NoHostAvailable
 from util import new_test_table
-
+from test.pylib.rest_client import ScyllaRESTAPIClient, inject_error
```

Comment:
I'll prepare now myself a patch that adds rest_api and scylla_inject_error to the cql-pytest framework, and let you review it.

Reply:
Can you please take a look and review https://github.com/scylladb/scylladb/pull/12264 and see if the `scylla_inject_error()` function I defined in that patch suites your needs? Thanks.

Reply:
> What I would have prefered to see is copying of scylla_inject_error() from test/alternator/util.py to test/cql-pytest/util.py. It's a really simple function, and doesn't need to use asyncio which nothing else in the test/cql-pytest directory uses today.

I'm not in favour of this. We can copy-paste this concrete function, but this reinforces the boundary between the 'test worlds', hampering code reuse. Pylib-world is actively evolving now, and this patch allows us to reuse its components in cql-pytest in the future.

The problem with importing `test.pylib.rest_client` seems easily fixed with a single line of `pytest.ini` config:
`pythonpath = ../..`

> Please also check that this test gets skipped on Cassandra (test/cql-pytest/run-cassandra) because Cassandra doesn't have our injection mechanism.

 I added `scylla_only`, thanks.

Reply:
I think that this should be discussed outside a specific PR adding a Scylla feature.

In general, I am troubled by the inflation of test/pylib. It's already 2000 lines, not including test.py itself which has a lot of pylib-specific code. All this for just 900 lines of test code that uses it in test_topology*/. Compare this to 800 lines of library code in test/cql-pytest (including test/cql-pytest/cassandra-tests/porting.py), used by 27,000 lines of code in test/cql-pytest.

I am not enthusiastic about a vast library of testing functions. I think this approach has done a lot of harm in the dtest world. They have a huge number of "convenience" functions which test developers are tempted to use even when much more direct approaches are available. There's often 5 different functions that do a similar thing, and when none of them do exactly what the test developer wants, he develops a 6th variant and adds it to the test library. The tests themselves become opaque because instead of being written using concepts that are being tested (like CQL requests) they use test functions which nobody is familiar with and might themselves be buggy (and this frequently happens).

I think that if we only have a small number of test convenience functions, duplicating them doesn't do much harm. Moreover when the duplication isn't exact (e.g., no need for async in a test suite that doesn't use that anywhere else), it's cleaner to just have a clean copy instead of weird hacks that make a single copy be useful for different test suites.

Finally, the "pylib" world not just has too little real-world experience (only 900 lines of test code that use it), it also has no experience at all with tests that are not cql. We have 17,000 lines of test code (700 tests) in test/alternator. We have test/redis, test/rest_api. These reuse *some* similar convenience functions, but other functions are completely different, because Alternator, Redis and CQL are not the same thing. The inject_error function we discuss here was actually written for Alternator first, and pylib copied the initial implementation from test/alternator. Would test/alternator use pylib too? It can't unless the pylib becomes less cql-specific. It will take some work to design a pylib-like library that will be useful for all these test subdirectories, and I'm not sure the effort is justified. The Alternator test "library" (test/alternator/conftest.py, test/alternator/util) is also just 500 lines used by 17,000 lines of test code, so reducing the duplication in this library is also of dubious benefit.

Reply:
@nyh if it's a general comment, it's fine, we can move it into an email thread, maybe find some alternatives/improvements. But what does it imply for this patch? There is no specific request in this comment, it makes it ambiguous. Meanwhile, the patch fixes a genuine bug and has not been moving forward for a while.

Reply:
If it wasn't clear already: I sent PR https://github.com/scylladb/scylladb/pull/12264 which adds `scylla_inject_error()` to cql-pytest, and it was committed. Please just use this in this patch instead of attempting to use code from pylib inside cql-pytest. It will also simplify your test patch to only contain your real test, and not mix it with framework improvements.